source/Octopus.Tentacle.Tests.Integration/WorkspaceCleanerTests.cs

```diff
@@ -0,0 +1,156 @@
+using Octopus.Tentacle.CommonTestUtils.Builders;
+using Octopus.Tentacle.Maintenance;
+using Octopus.Tentacle.Scripts;
+using Octopus.Tentacle.Tests.Integration.Support;
+using Octopus.Tentacle.Tests.Integration.Util;
+using Octopus.Tentacle.Tests.Integration.Util.Builders;
+using Octopus.Tentacle.Tests.Integration.Util.Builders.Decorators;
+
+namespace Octopus.Tentacle.Tests.Integration
+{
+    [IntegrationTestTimeout]
+    public class WorkspaceCleanerTests : IntegrationTest
+    {
+        [Test]
+        [TentacleConfigurations(testDefaultTentacleRuntimeOnly: true)]
+        public async Task WhenScriptServiceIsRunning_ThenWorkspaceIsNotDeleted(TentacleConfigurationTestCase tentacleConfigurationTestCase)
+        {
+            var cleanerDelay = TimeSpan.FromMilliseconds(500);
+            var deleteWorkspacesOlderThan = TimeSpan.FromMilliseconds(500);
+
+            var startScriptCommand = new StartScriptCommandV2Builder().WithScriptBody(b => b.Print("Hello")).Build();
+
+            var workspaceDirectory = string.Empty;
+            await using var clientAndTentacle = await tentacleConfigurationTestCase.CreateBuilder()
+                .WithTentacle(b =>
+                {
+                    b.WithRunTentacleEnvironmentVariable(WorkspaceCleanerConfiguration.CleanerDelayEnvironmentVariableName, cleanerDelay)
+                        .WithRunTentacleEnvironmentVariable(WorkspaceCleanerConfiguration.DeleteWorkspacesOlderThanTimeSpanEnvironmentVariableName, deleteWorkspacesOlderThan);
+                })
+                .WithTentacleServiceDecorator(new TentacleServiceDecoratorBuilder()
+                    .DecorateScriptServiceV2With(d =>
+                        d.BeforeCompleteScript(async (_, _) =>
+                            {
+                                Directory.Exists(workspaceDirectory).Should().BeTrue("Workspace should exist before we complete");
+
+                                await Task.Delay(1000, CancellationToken);
+
+                                Directory.Exists(workspaceDirectory).Should().BeTrue("Workspace should not have been cleaned up");
+                            })
+                            .Build())
+                    .Build())
+                .Build(CancellationToken);
+
+            workspaceDirectory = GetWorkspaceDirectoryPath(clientAndTentacle.RunningTentacle.HomeDirectory, startScriptCommand.ScriptTicket.TaskId);
+
+            await clientAndTentacle.TentacleClient.ExecuteScript(startScriptCommand, CancellationToken, null, new InMemoryLog());
+            
+            Directory.Exists(workspaceDirectory).Should().BeFalse("Workspace should be naturally cleaned up after completion");
+        }
+        
+        [Test]
+        [TentacleConfigurations(testDefaultTentacleRuntimeOnly: true)]
+        public async Task WhenTentacleClientCrashesAndTentacleNeverGotToldToComplete_ThenWorkspaceIsConsideredRunning_AndWillNotBeDeleted(TentacleConfigurationTestCase tentacleConfigurationTestCase)
```

Comment:
I nearly didn't put this test case in, but I thought it an interesting one that's at least worth looking at during the PR.

Reply:
This ended up being a problem. So we fixed the issue.